src/main/java/graphql/execution/async/AsyncExecutionStrategy.java

```diff
@@ -0,0 +1,199 @@
+package graphql.execution.async;
+
+import com.spotify.futures.CompletableFutures;
+import graphql.ExceptionWhileDataFetching;
+import graphql.ExecutionResult;
+import graphql.GraphQLException;
+import graphql.execution.ExecutionContext;
+import graphql.execution.ExecutionStrategy;
+import graphql.language.Field;
+import graphql.schema.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.stream.Collectors.toList;
+
+
+public final class AsyncExecutionStrategy extends ExecutionStrategy {
+
+    public static AsyncExecutionStrategy serial() {
+        return new AsyncExecutionStrategy(true);
+    }
+
+    public static AsyncExecutionStrategy parallel() {
+        return new AsyncExecutionStrategy(false);
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(AsyncExecutionStrategy.class);
+
+    private final boolean serial;
+
+    private AsyncExecutionStrategy(boolean serial) {
+        this.serial = serial;
+    }
+
+    @Override
+    public ExecutionResult execute(ExecutionContext executionContext, GraphQLObjectType parentType, Object source, Map<String, List<Field>> fields) {
+        try {
+            return executeAsync(executionContext, parentType, source, fields).toCompletableFuture().get();
+        } catch (Throwable e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public CompletionStage<ExecutionResult> executeAsync(ExecutionContext executionContext, GraphQLObjectType parentType, Object source, Map<String, List<Field>> fields) {
+        Map<String, Supplier<CompletionStage<ExecutionResult>>> executingFields = new LinkedHashMap<>();
+
+        for (String fieldName : fields.keySet()) {
```

Comment:
This for loop needs to be replaced with a construct that properly closes over the vars passed to `resolveFieldAsync`